src/ioctls/common/kvm_vec.rs

```diff
@@ -0,0 +1,549 @@
+    // must be contiguous. Since the entries are of type `KvmArray::Entry`
+    // we must be careful to convert the desired capacity of the `KvmVec`
+    // from `KvmArray::Entry` to `T` when reserving or releasing memory.
+    mem_allocator: Vec<T>,
+    // the number of elements of type `KvmArray::Entry` currently in the vec
+    len: usize,
+    // the capacity of the `KvmVec` measured in elements of type `KvmArray::Entry`
+    capacity: usize,
+}
+
+impl<T: Default + KvmArray> KvmVec<T> {
+    /// Get the capacity required by mem_allocator in order to hold
+    /// the provided number of  `KvmArray::Entry`
+    ///
+    fn kvm_vec_len_to_mem_allocator_len(kvm_vec_len: usize) -> usize {
+        let kvm_vec_size_in_bytes = size_of::<T>() + kvm_vec_len * size_of::<T::Entry>();
+        (kvm_vec_size_in_bytes + size_of::<T>() - 1) / size_of::<T>()
+    }
+
+    /// Get the number of elements of type `KvmArray::Entry` that fit
+    /// in a mem_allocator of provided len
+    ///
+    fn mem_allocator_len_to_kvm_vec_len(mem_allocator_len: usize) -> usize {
+        if mem_allocator_len == 0 {
+            return 0;
+        }
+
+        let array_size_in_bytes = (mem_allocator_len - 1) * size_of::<T>();
+        array_size_in_bytes / size_of::<T::Entry>()
+    }
+
+    /// Constructs a new KvmVec<T> that contains `num_elements` empty elements
+    /// of type `KvmArray::Entry`
+    ///
+    /// # Arguments
+    ///
+    /// * `num_elements` - The number of empty elements of type `KvmArray::Entry` in the initial `KvmVec`
+    ///
+    pub fn new(num_elements: usize) -> KvmVec<T> {
+        let required_mem_allocator_capacity =
+            KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(num_elements);
+
+        let mut mem_allocator = Vec::with_capacity(required_mem_allocator_capacity);
+        for _ in 0..required_mem_allocator_capacity {
+            mem_allocator.push(T::default())
+        }
+        mem_allocator[0].set_len(num_elements);
+
+        KvmVec {
+            mem_allocator,
+            len: num_elements,
+            capacity: num_elements,
+        }
+    }
+
+    /// Creates a new `KvmVec` structure based on a supplied vector of `KvmArray::Entry`.
+    ///
+    /// # Arguments
+    ///
+    /// * `entries` - The vector of `KvmArray::Entry` entries.
+    ///
+    pub fn from_entries(entries: &[T::Entry]) -> KvmVec<T> {
+        let mut kvm_vec = KvmVec::<T>::new(entries.len());
+
+        {
+            let kvm_vec_entries = kvm_vec.as_mut_kvm_struct().entries_mut();
+            // this is safe because the provided length is correct
+            let kvm_vec_entries_slice = unsafe { kvm_vec_entries.as_mut_slice(entries.len()) };
+            kvm_vec_entries_slice.copy_from_slice(entries);
+        }
+
+        kvm_vec
+    }
+
+    /// Get a reference to the actual KVM structure instance.
+    ///
+    pub fn as_kvm_struct(&self) -> &T {
```

Comment:
I think that, if it's fine to call the method below that returns a `*const T` pointer `as_ptr`, then this one can be called `as_ref`, and the following `as_mut`.